ADBDEV-7428: Add missing field initialization in wal record of bitmap index

mos65o2 opened the pull request:
Add missing field initialization in wal record of bitmap index (#1723)

In `_bitmap_write_bitmapwords_on_page()` the wal record (`xlrec_perpage` structure)
is initialized, which is used to replicate the bitmap index data to the mirror.
Initialization of `bm_next_blkno` was missed here.
After vacuum, this led to inconsistency and corruption of the bitmap index on
the mirror (the difference between wal record and the bitmap index page data
can be detected in `_bitmap_log_bitmapwords()`).

bandetto commented on src/test/regress/sql/bitmap_index.sql:
```diff
@@ -479,3 +479,33 @@ select * from bm_test where b = 1;
+\c - - - 7000
+drop table ao_table;
+-- start_ignore
+\! gpconfig -c hot_standby -v off
```
Maybe just reset the value to default?
```suggestion
\! gpconfig -r hot_standby
```

bandetto commented on src/test/regress/sql/bitmap_index.sql:
```diff
@@ -479,3 +479,33 @@ select * from bm_test where b = 1;
+create index bitmap_idx_ao on ao_table using bitmap(a);
+delete from ao_table where a < 8;
+
+\! psql postgres -c "select pg_sleep(5)" &
```
Maybe using isolation2 suite may be more convenient for this test?

https://github.com/arenadata/gpdb/blob/d2443f802c05d8c36d066cece00972ab1387db21/src/test/isolation2/sql_isolation_testcase.py#L856-L1024

mos65o2 replied:
I don't see much difference between these approaches. If I can find a way to reproduce the problem without `pg_sleep`, it might be worth doing.

bimboterminator1 replied:
it can't be reproduced without [parallel transaction](https://github.com/arenadata/gpdb/blob/513b8d1f123eacd1742172377f24efda6abc8c0e/src/backend/access/appendonly/appendonly_compaction.c#L591-L631), so the pg_sleep suits well here. 

Secondly, i suggest showing the result of patch via checking  checksums of corresponding pages ( equality of `SELECT checksum FROM page_header(get_raw_page('idx_bitmap_reindex_ao', 9));
`) at mirror and primary rather than via plain select.

Or, we could do that without cluster restart, for example, writing plpython function which compares 2 relfilenodes at mirror and at primary

mos65o2 replied:
It seems that testing in this case is not trivial. We have to reboot the cluster in hot_standby mode. The query with `select` indirectly checks the functionality of the patch. Other testing methods did not yield results.

bandetto commented on src/test/regress/sql/bitmap_index.sql:
```diff
@@ -479,3 +479,33 @@ select * from bm_test where b = 1;
+
+\! psql postgres -c "select pg_sleep(5)" &
+vacuum ao_table;
+-- start_ignore
+\! gpconfig -c hot_standby -v on
+\! gpstop -afr
+-- end_ignore
```
I think removing ignored blocks which directly affect the test for expected/ files gets rid of too much context. What do you think about keeping them intact?

bandetto commented on src/test/regress/sql/bitmap_index.sql:
```diff
@@ -479,3 +479,33 @@ select * from bm_test where b = 1;
+\c
+set enable_seqscan = off;
+select * from ao_table where a in (7, 8, 9, 10);
+-- mirror
```
Perhaps you meant:
```suggestion
-- standby
```

bandetto commented on src/test/regress/sql/bitmap_index.sql:
```diff
@@ -479,3 +479,33 @@ select * from bm_test where b = 1;
 
 -- clean up
 drop table bm_test;
+
```
Extra newline.
```suggestion
```

bimboterminator1 commented on src/backend/access/bitmap/bitmapinsert.c:
```diff
@@ -1495,6 +1495,7 @@ _bitmap_write_bitmapwords_on_page(Page bitmapPage, BMTIDBuffer *buf, int startWo
 		if (xlrec_perpage)
 		{
 			xlrec_perpage->bmp_last_tid = bitmapPageOpaque->bm_last_tid_location;
+			xlrec_perpage->bm_next_blkno = bitmapPageOpaque->bm_bitmap_next;
```
Why not just put the line
`xlrec_perpage->bm_next_blkno = InvalidBlockNumber;`
before the [condition](https://github.com/arenadata/gpdb/blob/513b8d1f123eacd1742172377f24efda6abc8c0e/src/backend/access/bitmap/bitmaputil.c#L1078)

mos65o2 replied:
It looks simple and ok. But are we always sure that if the condition is false, `xlrec_perpage->bm_next_blkno` should be InvalidBlockNumber? Now it is a part of copy of bitmapPage.

bimboterminator1 replied:
How do we choose the bitmap page which we are inserting the tid into? If we can insert in single page somewhere in the middle of bitmap vector, we should initialize from bimapPageOpaque, i agree

mos65o2 replied:
I think the first bitmapPage (opaque->bm_bitmap_next) can accept not only InvalidBlockNumber.
In case of a [new buffer](https://github.com/arenadata/gpdb/blob/513b8d1f123eacd1742172377f24efda6abc8c0e/src/backend/access/bitmap/bitmapinsert.c#L1290), we [initialize](https://github.com/arenadata/gpdb/blob/513b8d1f123eacd1742172377f24efda6abc8c0e/src/backend/access/bitmap/bitmapinsert.c#L1364) a new page, and here InvalidBlockNumber is explicitly [written](https://github.com/arenadata/gpdb/blob/513b8d1f123eacd1742172377f24efda6abc8c0e/src/backend/access/bitmap/bitmappages.c#L155).
But if we get an [existing buffer](https://github.com/arenadata/gpdb/blob/513b8d1f123eacd1742172377f24efda6abc8c0e/src/backend/access/bitmap/bitmapinsert.c#L1277) then we [do not initialize](https://github.com/arenadata/gpdb/blob/513b8d1f123eacd1742172377f24efda6abc8c0e/src/backend/access/bitmap/bitmapinsert.c#L1347) it.
The buffer eventually ends up in `_bitmap_log_bitmapwords` via [perpage_buffers](https://github.com/arenadata/gpdb/blob/513b8d1f123eacd1742172377f24efda6abc8c0e/src/backend/access/bitmap/bitmapinsert.c#L1312).

mos65o2 committed:
Add missing initialization of bm_next_blkno field (#1723)

In _bitmap_write_bitmapwords_on_page() the wal record (xlrec_perpage structure)
is initialized, which is used to replicate the bitmap index data to the mirror.
Initialization of bm_next_blkno was missed here.
After vacuum, this led to inconsistency and corruption of the bitmap index on
the mirror (the difference between wal record and the bitmap index page data
can be detected in _bitmap_log_bitmapwords()).
